collections-arrays/CompareByteArraysInCsharp/CompareByteArraysInCsharp/CompareByteArrays.cs

```diff
@@ -64,75 +64,102 @@ public bool CompareUsingForLoop(byte[] firstArray, byte[] secondArray)
+                        return false;
+                }
+
+                for (int i = 0; i < arrayLength; i++)
+                {
+                    if (pbtr1[i] != pbtr2[i])
+                        return false;
                 }
             }
 
             return true;
         }
 
+        private unsafe bool VectorEquality(byte* firstPointer, byte* secondPointer)
+        {
+            var firstVector = *(Vector<byte>*)firstPointer;
+            var secondVector = *(Vector<byte>*)secondPointer;
+
+            return Vector.EqualsAll(firstVector, secondVector);
+        }
+
         [ArgumentsSource(nameof(SampleByteArray))]
         [Benchmark]
-        public bool CompareUsingIStructuralEquatable(byte[] firstArray, byte[] secondArray)
+        public bool CompareUsingIStructuralEquatable(byte[] firstArray, byte[] secondArray, string arrayName)
         {
             if (firstArray == null || secondArray == null || firstArray.Length != secondArray.Length)
                 return false;
 
             return StructuralComparisons.StructuralEqualityComparer.Equals(firstArray, secondArray);
         }
 
-        [DllImport("msvcrt.dll", CallingConvention = CallingConvention.Cdecl)]
-        private static extern int memcmp(byte[] firstArray, byte[] secondArray, int size);
+#if NET
+        [DllImport("msvcrt", CallingConvention = CallingConvention.Cdecl, SetLastError = true)]
+        private static extern int memcmp(byte[] firstArray, byte[] secondArray, long size);
+#else
+        [DllImport("libc", CallingConvention = CallingConvention.Cdecl, SetLastError = true)]
+        private static extern int memcmp(byte[] firstArray, byte[] secondArray, ulong size);
+#endif
+
         [ArgumentsSource(nameof(SampleByteArray))]
         [Benchmark]
-        public bool CompareUsingPInvoke(byte[] firstArray, byte[] secondArray)
+        public bool CompareUsingPInvoke(byte[] firstArray, byte[] secondArray, string arrayName)
         {
             if (firstArray == null || secondArray == null || firstArray.Length != secondArray.Length)
+            {
                 return false;
+            }
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return memcmp(firstArray, secondArray, firstArray.Length) == 0;
+            }
 
-            return memcmp(firstArray, secondArray, firstArray.Length) == 0;
+            else
+            {
+                return memcmp(firstArray, secondArray, firstArray.Length) == 0;
+            }
         }
 
-        private byte[] GenerateRandomArray(int size) 
+        private byte[] GenerateOrderedArray(bool firstElement, bool middleElement, int size)
         {
-            var byteArray = new byte[size];
+            var intArray = new int[size];
+            var byteArray = new byte[size * sizeof(int)];
+
+            for (int i = 0; i < size; i++)
+            {
+                if (firstElement == true && i == 0)
```

Comment:
You can simplify this with just if (firstElement && i == 0). There is no need to check == for boolean values. The same applies for the line with middleElement.

This whole method could be simplified a little bit (in my opinion, feel free to disagree). You could just use the single byte array declaration. Then you have the option of doing it as an unsafe block, fixing the array and then casting the fixed pointer to an int pointer. Or you could use the MemoryMarshal.GetArrayDataReference method to take a reference to the array (no need for an unsafe block). You can then use the Unsafe.As to get an int reference to the byte array. You just have to be careful with your index value, but you would then be looking at the byte array as an int array so it would remove the need for the extra array allocation as well as the copy. Again, that is just my opinion.

Reply:
@jshergal, that's a great idea. I will be pushing another version shortly.